File: zad1/zad1.cpp
#include <iostream>
#include <complex>
#include <stdio.h>

using namespace std;

#define PI atan(1)*4


complex<double> func(int n, double delta, double beta, double d, double teta) {
	complex<double> sum;
	sum.real(0.0);
	sum.imag(0.0);

	complex<double> num;

	double fi = delta + d * beta * cos(teta);

	for (int k = 0; k < n; k++) {
		num.real(cos(-k * fi));
		num.imag(sin(-k * fi));
		sum += num;
	}

	return sum;
}

int main() {
	FILE* out;
	fopen_s(&out, "complex.txt", "w");

	int n = 5;
	double beta = 20 * PI;
	double d = 1 / 20;
	double teta = PI / 4;

	complex<double> num;
	for (double delta = 0.0; delta <= 2 * PI; delta += 0.05) {
		num = func(n, delta, beta, d, teta);
		fprintf(out, "%.5f %.5f\n", delta, sqrt(pow(num.real(), 2) + pow(num.imag(), 2)));
	}
	fclose(out);
}

File: zad1/zad1.py
import numpy as np
from scipy.optimize import minimize
import matplotlib.pyplot as plt
import math

##### READ DATA ##########
file = open('complex.txt', 'r')
delta = []
f = []
for line in file:
    s = line.split()
    if(len(s) != 0):
        delta.append(float(s[0]))
        f.append(float(s[1]))
file.close()

##### DRAW PLOT ##########
plt.figure()
plt.grid(b = 'true')
plt.plot(delta, f)
plt.xlabel('delta')
plt.ylabel('|F(delta)|')

##### DEFINE FUNCTION ##########
def fun(params, constants):
    delta = params[0]
    n, beta, d, teta = constants
    sum = [0, 0]
    fi = delta + d * beta * math.cos(teta)
    for k in range(int(n)):
        sum[0] = sum[0] + math.cos((-k) * fi)
        sum[1] = sum[1] + math.sin((-k) * fi)
    return math.sqrt(sum[0]**2 + sum[1]**2)

##### MINIMIZE FUNCTION ##########
constants = [5, 20*math.pi, 0.05, math.pi/4]
init_guess = [4]
res = minimize(lambda p,c: -fun(p,c), init_guess, args = constants, method='Nelder-Mead', options={'ftol': 1e-8, 'disp': True, 'maxfev': 1e4})
xmin = np.array(res.x)
print("max = ", xmin[0])

##### DRAW MAXIMUM ##########
plt.scatter(xmin[0], fun([xmin[0]], constants), facecolor='red')